controller/array_action/array_mediator_svc.py

```diff
@@ -1292,3 +1309,58 @@ def _get_host_name_if_equal(self, nvme_host, fc_host, iscsi_host):
         if len(unique_names) == 1:
             return unique_names.pop()
         return None
+
+    def _addsnapshot(self, name, source_volume_id, pool):
+        try:
+            return self.client.svctask.addsnapshot(name=name, volumes=source_volume_id, pool=pool)
+        except (svc_errors.CommandExecutionError, CLIFailureError) as ex:
+            if is_warning_message(ex.my_message):
+                logger.warning("exception encountered while creating snapshot '{}': {}".format(name,
+                                                                                               ex.my_message))
+            else:
+                logger.error("cannot create snapshot {0}, Reason is: {1}".format(name, ex))
+                if OBJ_ALREADY_EXIST in ex.my_message:
+                    raise array_errors.SnapshotAlreadyExists(name, self.endpoint)
+                if NAME_NOT_EXIST_OR_MEET_RULES in ex.my_message:
+                    raise array_errors.PoolDoesNotExist(pool, self.endpoint)
+                if NOT_ENOUGH_EXTENTS_IN_POOL_CREATE in ex.my_message:
+                    raise array_errors.NotEnoughSpaceInPool(id_or_name=pool)
+                if NOT_CHILD_POOL in ex.my_message:
+                    raise array_errors.PoolDoesNotExist(pool, self.endpoint)
+                if any(msg_id in ex.my_message for msg_id in (NON_ASCII_CHARS, INVALID_NAME, TOO_MANY_CHARS)):
+                    raise array_errors.IllegalObjectName(ex.my_message)
```

Comment:
are all of these relevant for flashcopy 2.0?
if so, could we wrap them in a `_create_object(create_func, **kwargs)` method?

Reply:
or maybe a `@create_object` decorator

Reply:
CSI-4261